README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Thomas Reufer <thomas@reufer.ch>

Project's homepage:          https://github.com/reufer/rpihddevice

Latest version available at: https://github.com/reufer/rpihddevice

See the file COPYING for license information.

Description:

  VDR HD output device for Raspberry Pi. The plugin makes use of the Raspberry
  Pi's VideoCore GPU and provides a lightweight implementation for a VDR output
  device.

Features:

  - MPEG-2 and H264 high-profile video codec up to 1080p30
  - MPEG-1 Layer II, (E)AC-3, AAC and DTS audio codec at 32kHz, 44.1kHz or 48kHz
    with 2.0 (Stereo) or 5.1 channels
  - HDMI multi channel LPCM audio output
  - HDMI digital audio pass-through
  - Analog stereo audio output
  - Box (letter-box/pillar-box), Crop and Stretch video display modes
  - True color OSD with GPU support
  - Video scaling and grabbing support

Requirements:

  - libavcodec, libavformat and libavutil for audio decoding, provided by ffmpeg
    or libav
  - libswresample when using ffmpeg-1.2 or newer
  - libavresample when using libav-9 or newer
  - freetype2 for GPU accelerated text drawing
  - valid MPEG-2 license when watching MPEG-2 streams
  - Raspberry Pi userland libraries: https://github.com/raspberrypi/userland
  - Raspberry Pi firmware version of 2015/01/18 or newer

Install:

  Get the source code either as archive or with git and compile like any other
  VDR plugin:

  $ cd /usr/src/vdr/PLUGINS/src
  $ git clone git://projects.vdr-developer.org/vdr-plugin-rpihddevice.git rpihddevice
  $ cd rpihddevice
  $ make
  $ make install

  If you want to link the plugin against a specific version of ffmpeg/libav, set
  EXT_LIBAV accordingly when compiling the plugin:

  $ make EXT_LIBAV=/usr/src/ffmpeg-1.2.6

Usage:

  To start the plugin, just add '-P rpihddevice' to the VDR command line.

  The plugin simply adds two new dispmanx layers on top of the framebuffer, one
  for video and one for the OSD. The plugin does not clear the current console
  or change any video mode settings. So it's the user's choice, what's being
  displayed when no video is shown, e.g. during channel switches or for radio
  channels.

Options:

  -d, --disable-osd  Disables creation of OSD layer and prevents the plugin of
                     allocating any OSD related resources. If set, VDR's dummy
                     OSD is used, when selecting rpihddevice as primary device.
  -v, --video-layer  Specify the dispmanx layer for video (default 0)
  -o, --osd-layer    Specify the dispmanx layer for OSD (default 2)
      --display      display used for output:
                     0: default display (default)
                     4: LCD
                     5: TV/HDMI
                     6: non-default display

Plugin-Setup:

  Resolution: Set video resolution. Possible values are: "default",
  "follow video", "720x480", "720x576", "1280x720", "1920x1080"

  Frame Rate: Set video frame rate. Possible values are: "default",
  "follow video", "24p", "25p", "30p", "50i", "50p", "60i", "60p"

  When set to "default", the resolution/frame rate will not be changed and the
  plugin keeps the current setting of the framebuffer, normally set by the mode
  number in /boot/config.txt or changed with 'tvservice'. To let the plugin
  automatically set a value matching the current video stream, choose
  "follow video". In general, video setting are only applied if both, resolution
  and frame rate, fits to a video mode supported by the connected device,
  indicated by its EDID.

  EDID information can by overridden with various settings in /boot/config.txt,
  see the official documentation for further information:
  https://www.raspberrypi.org/documentation/configuration/config-txt.md

  Video Framing: Determines how the video frame is displayed on the screen in
  case the aspect ratio differs. "box" and "cut" will preserve the video's
  aspect ratio, while "box" (often called "letter box", however "pillar box" is
  used to show 4:3 videos on a wide screen) will fit the image on the
  screen by adding transparent borders. On the other hand, "cut" is cropping
  away the overlapping part of the video, so the entire display area is filled.
  When setting to "stretch", the videos' aspect ratio is adapted to the screen
  and the resulting image might appear distorted.

  Audio Port: Set the audio output port to "analog" or "HDMI". When set to
  analog out, multi channel audio is sampled down to stereo.

  Digital Audio Format: Specify the audio format when using the HDMI port as
  output. If set to "pass through", (E)AC-3 and DTS audio can be decoded by the
  connected HDMI device, if EDID indicates specific codec support. For local
  decoding, select "mutli channel PCM" or "Stereo PCM" if additional stereo
  dowmix of mutli channel audio is desired.

  Use GPU accelerated OSD: Use GPU capabilities to draw the on screen display.
  Disable acceleration in case of OSD problems to use VDR's internal rendering
  and report error to the author.



HISTORY
VDR Plugin 'rpihddevice' Revision History
-----------------------------------------

2022-12-26: Version 1.0.6
-------------------------
- fixed:
  - project URL and whitespaces in README (thanks to Martin Dummer)
  - overflows reported by GCC -fsanitize=undefined (thanks to Marko Mäkelä)
  - hangs and crashes on shutdown due to OMX deinitialization (thanks to Marko Mäkelä)
  - potential memory leak in cRpiAudioDecoder::m_ptsQueue (thanks to Marko Mäkelä)
  - SIGFPE (division by 0) on startup if screen pixelHeight is 0 (thanks to Marko Mäkelä) 
  - no picture after switching channels (thanks to Marko Mäkelä)
  - link -lbrcmEGL before -lbrcmGLESv2 (thanks to Marko Mäkelä)

2019-06-16: Version 1.0.5
-------------------------
- fixed:
  - drawing of empty strings on high level OSD (thanks to Klaus Schmidinger)
  - compilation for Raspbian Stretch (thanks to Klaus Schmidinger)
  - compilation for ffmpeg-4 (thanks to Stefan Schallenberg)

2018-07-16: Version 1.0.4
-------------------------
- new:
  - updated Hungarian translations (thanks to Füley István)
  - make use of advanced deinterlacer configurable
  - add debug option to log number of executed OpenVG commands and flushes
  - set OMX clock pre-roll to 250ms for live TV (transfer mode)
- fixed:
  - reset video format settings on pixel aspect ratio change 
  - always resample audio with less than 2 and  more than 6 channels
  - fixed compilation with GCC-6
  - implement proper handling of display and pixel aspect ratios
  - fixed vertical text position

2016-04-23: Version 1.0.3
-------------------------
- new:
  - re-enable advanced deinterlacer for SD streams, requires recent firmware
- fixed:
  - don't ignore audio frames for slow trick speeds to keep clock in sync
  
2016-04-04: Version 1.0.2
-------------------------
- fixed:
  - send proper end of sequence packet after playing single video frame

2016-03-29: Version 1.0.1
-------------------------
- new:
  - set device name
  - add option for dual display usage, see README
  - experimental support for AAC-LATM, use "make ENABLE_AAC_LATM=1" to enable
- fixed:
  - fixed HDMI channel mapping and stream type (reported by Rüdiger Follmann)
  - fixed drawing of cached images with high level OSD
  - only cache images in GPU when high level OSD is active
  - added proper GPL header to every source file (suggested by Tobias Grimm)
  - wait until dispmanx element is removed after OSD reset
  - proper release default surface after OSD reset
  - don't ignore anti alias option for scaled bitmap drawing
  - ignore pts from erroneous packets to avoid pts jumps on broken streams
  - use omxplayer's parameters for deinterlacer

2015-10-18: Version 1.0.0
-------------------------
- new:
  - updated Hungarian translations (thanks to Füley István)
  - updated Finnish translations (thanks to Rolf Ahrenberg)
  - added Italian translations (thanks to Gerlando Falauto)
  - added French translations (thanks to Cyril Jaquier)
  - added plugin options to specify video and OSD layers
- fixed:
  - don't crop OSD images to pixmap size to allow scaled drawing
  - treat zero as valid PTS value
  - perform save / restore region direct on pixmaps for accelerated OSDs
  - wait for the OVG thread to be ready when creating an accelerated OSD
  - set field sync when output interlaced material at interlaced display mode

2015-04-29: Version 0.1.0
-------------------------
- new:
  - reworked clock stretching in live mode based on buffer usage
  - added command line argument to disable OSD
  - combined digital audio format options to one single setup option
  - added font kerning
  - support for GPU accelerated pixmaps
- fixed:
  - skip audio packets when replaying non-radio recordings in trick modes
  - fixed displaying of current position when changing replay speed
  - fixed PTS wrap around (reported by Klaus Schmidinger)
  - increased audio decoder thread priority
  - don't depend on multi channel PCM support for digital audio pass-through
  - improved video frame rate detection to be more tolerant to inaccurate values
  - adapted cOvgRawOsd::Flush() to new cOsd::RenderPixmaps() of vdr-2.1.10
  - fixed calculation of string height (reported by Klaus Schmidinger)
  - additionally set number of in/out channels when setting up audio resampler
  - don't clear audio PTS when parser shrinks the packet buffer

2015-01-21: Version 0.0.11
--------------------------
- new:
  - reworked audio decoder, increase buffering of audio samples
  - reworked omxdevice and improved clock stretching for transfer mode
  - added Hungarian translations (thanks to Füley István)
  - updated Finnish translations and fixed tr() usage (thanks to Rolf Ahrenberg)
  - use fast deinterlacer for HD streams
  - added high level OSD with GPU support
- fixed:
  - start H.264 video with valid frame only, requires firmware from 2015/01/18 
    or newer

2014-11-02: Version 0.0.10
--------------------------
- new:
  - added DTS audio codec support
  - add setup options to control video mode and its behaviour on stream changes
  - set HDMI speaker layout to fix channel mapping for multi channel PCM output
  - support building against external ffmpeg/libav by setting EXT_LIBAV
  - support for >=ffmpeg-1.2 and >=libav-0.8 with resampling
  - setup option for video framing mode in case of incompatible aspect ratio
  - redirect ffmpeg messages to plugin/VDR logs
  - added cppcheck as Makefile target (suggested by Rolf Ahrenberg)
  - reworked setup parameter handling
  - implemented proper command queue for OVG-OSD
- fixed:
  - increased latency target for live mode to 500ms
  - reworked OVG image processing for low level OSD
  - upmix mono audio channels to stereo when using PCM output on HDMI
  - skip non-video packets in StillPicture(), fixes artifacts in PES recordings
  - skip audio substream header for PES recordings with AC3 audio track
  - reworked audio decoding loop and set correct dstSize before resampling

setup.c
/*
 * rpihddevice - VDR HD output device for Raspberry Pi
 * Copyright (C) 2014, 2015, 2016 Thomas Reufer
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include "setup.h"
#include "display.h"
#include "ovgosd.h"

#include <vdr/tools.h>
#include <vdr/menuitems.h>

#include <getopt.h>

#include <bcm_host.h>
#include "interface/vchiq_arm/vchiq_if.h"
#include "interface/vmcs_host/vc_tvservice.h"

/* ------------------------------------------------------------------------- */

class cRpiSetupPage : public cMenuSetupPage
{

public:

	cRpiSetupPage(
			cRpiSetup::AudioParameters audio,
			cRpiSetup::VideoParameters video,
			cRpiSetup::OsdParameters osd) :

		m_audio(audio),
		m_video(video),
		m_osd(osd)
	{
		m_audioPort[0] = tr("analog");
		m_audioPort[1] = tr("HDMI");

		m_audioFormat[0] = tr("pass through");
		m_audioFormat[1] = tr("multi channel PCM");
		m_audioFormat[2] = tr("stereo PCM");

		m_videoFraming[0] = tr("box");
		m_videoFraming[1] = tr("crop");
		m_videoFraming[2] = tr("stretch");

		m_videoResolution[0] = tr("default");
		m_videoResolution[1] = tr("follow video");
		m_videoResolution[2] = "720x480 (4:3)";
		m_videoResolution[3] = "720x480 (16:9)";
		m_videoResolution[4] = "720x576 (4:3)";
		m_videoResolution[5] = "720x576 (16:9)";
		m_videoResolution[6] = "1280x720";
		m_videoResolution[7] = "1920x1080";

		m_videoFrameRate[0] = tr("default");
		m_videoFrameRate[1] = tr("follow video");
		m_videoFrameRate[2] = "24p";
		m_videoFrameRate[3] = "25p";
		m_videoFrameRate[4] = "30p";
		m_videoFrameRate[5] = "50i";
		m_videoFrameRate[6] = "50p";
		m_videoFrameRate[7] = "60i";
		m_videoFrameRate[8] = "60p";

		m_useAdvancedDeinterlacer[0] = trVDR("no");
		m_useAdvancedDeinterlacer[1] = tr("for SD video only");
		m_useAdvancedDeinterlacer[2] = tr("always");

		Setup();
	}

	eOSState ProcessKey(eKeys Key)
	{
		int newAudioPort = m_audio.port;
		eOSState state = cMenuSetupPage::ProcessKey(Key);

		if (Key != kNone)
		{
			if (newAudioPort != m_audio.port)
				Setup();
		}

		return state;
	}

protected:

	virtual void Store(void)
	{
		SetupStore("AudioPort", m_audio.port);
		SetupStore("AudioFormat", m_audio.format);

		SetupStore("VideoFraming", m_video.framing);
		SetupStore("Resolution", m_video.resolution);
		SetupStore("FrameRate", m_video.frameRate);
		SetupStore("AdvancedDeinterlacer", m_video.advancedDeinterlacer);

		SetupStore("AcceleratedOsd", m_osd.accelerated);

		cRpiSetup::GetInstance()->Set(m_audio, m_video, m_osd);
}

private:

	void Setup(void)
	{
		int current = Current();
		Clear();

		if (!cRpiDisplay::IsFixedMode())
		{
			Add(new cMenuEditStraItem(
				tr("Resolution"), &m_video.resolution, 8, m_videoResolution));

			Add(new cMenuEditStraItem(
				tr("Frame Rate"), &m_video.frameRate, 9, m_videoFrameRate));
		}
		if (cRpiDisplay::IsProgressive())
			Add(new cMenuEditStraItem(
					tr("Use Advanced Deinterlacer"),
					&m_video.advancedDeinterlacer, 3,
					m_useAdvancedDeinterlacer));

		Add(new cMenuEditStraItem(
				tr("Video Framing"), &m_video.framing, 3, m_videoFraming));

		Add(new cMenuEditStraItem(
				tr("Audio Port"), &m_audio.port, 2, m_audioPort));

		if (m_audio.port == 1)
		{
			Add(new cMenuEditStraItem(tr("Digital Audio Format"),
					&m_audio.format, 3, m_audioFormat));
		}

		Add(new cMenuEditBoolItem(
				tr("Use GPU accelerated OSD"), &m_osd.accelerated));

		SetCurrent(Get(current));
		Display();
	}

	cRpiSetup::AudioParameters m_audio;
	cRpiSetup::VideoParameters m_video;
	cRpiSetup::OsdParameters   m_osd;

	const char *m_audioPort[2];
	const char *m_audioFormat[3];
	const char *m_videoFraming[3];
	const char *m_videoResolution[8];
	const char *m_videoFrameRate[9];
	const char *m_useAdvancedDeinterlacer[3];
};

/* ------------------------------------------------------------------------- */

cRpiSetup* cRpiSetup::s_instance = 0;

cRpiSetup* cRpiSetup::GetInstance(void)
{
	if (!s_instance)
		s_instance = new cRpiSetup();

	return s_instance;
}

void cRpiSetup::DropInstance(void)
{
	delete s_instance;
	s_instance = 0;

	bcm_host_deinit();
}

bool cRpiSetup::HwInit(void)
{
	cRpiSetup* instance = GetInstance();
	if (!instance)
		return false;

	bcm_host_init();

	if (!vc_gencmd_send("codec_enabled MPG2"))
	{
		char buffer[1024];
		if (!vc_gencmd_read_response(buffer, sizeof(buffer)))
		{
			if (!strcasecmp(buffer,"MPG2=enabled"))
				GetInstance()->m_mpeg2Enabled = true;
		}
	}

	int width, height;
	if (!cRpiDisplay::GetSize(width, height))
	{
		ILOG("HwInit() done, display size is %dx%d", width, height);
	}
	else
		ELOG("failed to get video port information!");

	return true;
}

void cRpiSetup::SetAudioSetupChangedCallback(void (*callback)(void*), void* data)
{
	GetInstance()->m_onAudioSetupChanged = callback;
	GetInstance()->m_onAudioSetupChangedData = data;
}

void cRpiSetup::SetVideoSetupChangedCallback(void (*callback)(void*), void* data)
{
	GetInstance()->m_onVideoSetupChanged = callback;
	GetInstance()->m_onVideoSetupChangedData = data;
}

bool cRpiSetup::IsAudioFormatSupported(cAudioCodec::eCodec codec,
		int channels, int samplingRate)
{
	// MPEG-1 layer 2 audio pass-through not supported by audio render
	// and AAC audio pass-through not yet working
	if (codec == cAudioCodec::eMPG || codec == cAudioCodec::eAAC)
		return false;

	if (channels < 2 || channels > 6)
		return false;

	switch (GetAudioFormat())
	{
	case cAudioFormat::ePassThrough:
		return (vc_tv_hdmi_audio_supported(
					codec == cAudioCodec::eMPG  ? EDID_AudioFormat_eMPEG1 :
					codec == cAudioCodec::eAC3  ? EDID_AudioFormat_eAC3   :
					codec == cAudioCodec::eEAC3 ? EDID_AudioFormat_eEAC3  :
					codec == cAudioCodec::eAAC  ? EDID_AudioFormat_eAAC   :
					codec == cAudioCodec::eDTS  ? EDID_AudioFormat_eDTS   :
							EDID_AudioFormat_ePCM, channels,
					samplingRate ==  32000 ? EDID_AudioSampleRate_e32KHz  :
					samplingRate ==  44100 ? EDID_AudioSampleRate_e44KHz  :
					samplingRate ==  88200 ? EDID_AudioSampleRate_e88KHz  :
					samplingRate ==  96000 ? EDID_AudioSampleRate_e96KHz  :
					samplingRate == 176000 ? EDID_AudioSampleRate_e176KHz :
					samplingRate == 192000 ? EDID_AudioSampleRate_e192KHz :
							EDID_AudioSampleRate_e48KHz,
							EDID_AudioSampleSize_16bit) == 0);

	case cAudioFormat::eMultiChannelPCM:
		return codec == cAudioCodec::ePCM;

	default:
	case cAudioFormat::eStereoPCM:
		return codec == cAudioCodec::ePCM && channels == 2;
	}
}

void cRpiSetup::SetHDMIChannelMapping(bool passthrough, int channels)
{
	char command[80], response[80];

	sprintf(command, "hdmi_stream_channels %d", passthrough ? 1 : 0);
	vc_gencmd(response, sizeof(response), command);

	uint32_t channel_map = 0;

	if (!passthrough && channels > 0 && channels <= 6)
	{
		const unsigned char ch_mapping[6][8] =
		{
			{ 0, 0, 0, 0, 0, 0, 0, 0 }, // not supported
			{ 1, 2, 0, 0, 0, 0, 0, 0 }, // 2.0
			{ 1, 2, 4, 0, 0, 0, 0, 0 }, // 2.1
			{ 0, 0, 0, 0, 0, 0, 0, 0 }, // not supported
			{ 0, 0, 0, 0, 0, 0, 0, 0 }, // not supported
			{ 1, 2, 4, 3, 5, 6, 0, 0 }, // 5.1
		};

		// speaker layout according CEA 861, Table 28: Audio InfoFrame, byte 4
		const unsigned char cea_map[] =
		{
			0xff,	// not supported
			0x00,	// 2.0
			0x01,	// 2.1
			0xff,	// not supported
			0xff,	// not supported
			0x0b	// 5.1
		};

		for (int ch = 0; ch < channels; ch++)
			if (ch_mapping[channels - 1][ch])
				channel_map |= (ch_mapping[channels - 1][ch] - 1) << (3 * ch);

		channel_map |= cea_map[channels - 1] << 24;
	}

	sprintf(command, "hdmi_channel_map 0x%08x", channel_map);
	vc_gencmd(response, sizeof(response), command);
}

cMenuSetupPage* cRpiSetup::GetSetupPage(void)
{
	return new cRpiSetupPage(m_audio, m_video, m_osd);
}

bool cRpiSetup::Parse(const char *name, const char *value)
{
	if (!strcasecmp(name, "AudioPort"))
		m_audio.port = atoi(value);
	else if (!strcasecmp(name, "AudioFormat"))
		m_audio.format = atoi(value);
	else if (!strcasecmp(name, "VideoFraming"))
		m_video.framing = atoi(value);
	else if (!strcasecmp(name, "Resolution"))
		m_video.resolution = atoi(value);
	else if (!strcasecmp(name, "FrameRate"))
		m_video.frameRate = atoi(value);
	else if (!strcasecmp(name, "AdvancedDeinterlacer"))
		m_video.advancedDeinterlacer = atoi(value);
	else if (!strcasecmp(name, "AcceleratedOsd"))
		m_osd.accelerated = atoi(value);
	else return false;

	return true;
}

void cRpiSetup::Set(AudioParameters audio, VideoParameters video,
		OsdParameters osd)
{
	if (audio != m_audio)
	{
		m_audio = audio;
		if (m_onAudioSetupChanged)
			m_onAudioSetupChanged(m_onAudioSetupChangedData);
	}

	if (video != m_video)
	{
		m_video = video;
		if (m_onVideoSetupChanged)
			m_onVideoSetupChanged(m_onVideoSetupChangedData);
	}

	if (osd != m_osd)
	{
		m_osd = osd;
		cRpiOsdProvider::ResetOsd(false);
	}
}

bool cRpiSetup::ProcessArgs(int argc, char *argv[])
{
	const int cDisplayOpt = 0x100;
	static struct option long_options[] = {
			{ "disable-osd", no_argument,       NULL, 'd'         },
			{ "display",     required_argument, NULL, cDisplayOpt },
			{ "video-layer", required_argument, NULL, 'v'         },
			{ "osd-layer",   required_argument, NULL, 'o'         },
			{ 0, 0, 0, 0 }
	};
	int c;
	while ((c = getopt_long(argc, argv, "do:v:", long_options, NULL)) != -1)
	{
		switch (c)
		{
		case 'd':
			m_plugin.hasOsd = false;
			break;
		case 'o':
			m_plugin.osdLayer = atoi(optarg);
			break;
		case 'v':
			m_plugin.videoLayer = atoi(optarg);
			break;
		case cDisplayOpt:
		{
			int d = atoi(optarg);
			switch (d)
			{
			case 0:
			case 4:
			case 5:
			case 6:
				m_plugin.display = d;
				break;
			default:
				ELOG("invalid device id (%d), using default display!", d);
				break;
			}
		}
			break;
		default:
			return false;
		}
	}
	DBG("dispmanx layers: video=%d, osd=%d (%s), display=%d",
			m_plugin.videoLayer, m_plugin.osdLayer,
			m_plugin.hasOsd ? "enabled" : "disabled", m_plugin.display);

	return true;
}

const char *cRpiSetup::CommandLineHelp(void)
{
	return	"  -d,       --disable-osd  disable OSD\n"
			"  -v,       --video-layer  dispmanx layer for video (default 0)\n"
			"  -o,       --osd-layer    dispmanx layer for OSD (default 2)\n"
			"            --display      display used for output:\n"
			"                           0: default display (default)\n"
			"                           4: LCD\n"
			"                           5: TV/HDMI\n"
			"                           6: non-default display\n";
}
